### ISSUE TITLE
Trouble with email timeout

### ISSUE THREAD

**Issue by S0larflare** [gophish/gophish#608]
So a bit of background. An issue I came across when making a sending delay was that if the dialer was open for 10 minutes or if there wasn't anything sent for a few minutes, the next email to be sent would fail. Obviously, this is far from ideal. I came up with a hack to just renew the dialler all the time, but I felt that was probably an awful way of approaching the issue. Looking in GoMail, I couldn't see a retry that would be usable, so thought that catching the error which would be generated in either of these two scenarios, and recreating a new dialler in this instance would be the best way to move forward. So, with that, I wrote this in right after the     err = gomail.Send(s, e)      line in worker.go, line 211:

```
//if there is a connection timeout or broken pipe, create a new connection and try sending again...
                if ((err != nil && strings.Contains(err.Error(), "broken pipe")) || (err != nil && strings.Contains(err.Error(), "Connection timed out"))) {
        d = gomail.NewDialer(hp[0], port, c.SMTP.Username, c.SMTP.Password)
        d.TLSConfig = &tls.Config{
                ServerName:         c.SMTP.Host,
                InsecureSkipVerify: c.SMTP.IgnoreCertErrors,
        }
        hostname, err := os.Hostname()
        if err != nil {
                Logger.Println(err)
                hostname = "localhost"
        }
        d.LocalName = hostname

s, err = d.Dial()
         // Short circuit if we have an err
         // However, we still need to update each target
         if err != nil {
                 Logger.Println(err)
                 for _, t := range c.Results {
                         es := struct {
                                 Error string `json:"error"`
                         }{
                                Error: err.Error(),
                         }
                         ej, err := json.Marshal(es)
                         if err != nil {
                                 Logger.Println(err)
                         }
                         err = t.UpdateStatus(models.ERROR)
                         if err != nil {
                                 Logger.Println(err)
                         }
                         err = c.AddEvent(models.Event{Email: t.Email, Message: models.EVENT_SENDING_ERROR, Details: string(ej)})
                         if err != nil {
                                 Logger.Println(err)
                         }
                 }
         }
       err = gomail.Send(s, e)
}
//end of retry
```

Now, from a functional point of view, this works. If the email fails at first, it sends the second time, which is great. The problem is that the next line starts the normal error checking...and it doesn't pick up that err should now be null (if the err = gomail.Send(s, e) that has just run worked properly) and so writes it as an error into the database. I'm sure this is something to do with scoping in golang to do with errors, but I don't seem to be able to work it out at all. Is this some fundemental misunderstanding of what is going on here?

**Comment by jordan-wright** [comment]
Hey @S0larflare sorry for the delay on this one. I think this stems from the underlying design issues in our email worker. I'm hoping to rearchitect this piece soon that should alleviate quite a few of the issues we're seeing around email retries with the added benefit of making things like random delays, etc. easier.

**Comment by jordan-wright** [comment]
This should be fixed as part of the refactor from #878. If you could give it a shot and let me know if it helps, I'd appreciate it!
